commit 3a1261805940d0ff1dbbb9c705dddbc018c0423f
Author: Chuck Lever <chuck.lever@oracle.com>
Date:   Tue Aug 3 15:55:58 2021 -0400

    SUNRPC: Server-side disconnect injection
    
    Disconnect injection stress-tests the ability for both client and
    server implementations to behave resiliently in the face of network
    instability.
    
    A file called /sys/kernel/debug/fail_sunrpc/ignore-server-disconnect
    enables administrators to turn off server-side disconnect injection
    while allowing other types of sunrpc errors to be injected. The
    default setting is that server-side disconnect injection is enabled
    (ignore=false).
    
    Signed-off-by: Chuck Lever <chuck.lever@oracle.com>

diff --git a/net/sunrpc/svc.c b/net/sunrpc/svc.c
index 5aa263326b6a..bfcbaf7b3822 100644
--- a/net/sunrpc/svc.c
+++ b/net/sunrpc/svc.c
@@ -31,6 +31,8 @@
 
 #include <trace/events/sunrpc.h>
 
+#include "fail.h"
+
 #define RPCDBG_FACILITY	RPCDBG_SVCDSP
 
 static void svc_unregister(const struct svc_serv *serv, struct net *net);
@@ -1524,6 +1526,12 @@ svc_process(struct svc_rqst *rqstp)
 	struct svc_serv		*serv = rqstp->rq_server;
 	u32			dir;
 
+#if IS_ENABLED(CONFIG_FAIL_SUNRPC)
+	if (!fail_sunrpc.ignore_server_disconnect &&
+	    should_fail(&fail_sunrpc.attr, 1))
+		svc_xprt_deferred_close(rqstp->rq_xprt);
+#endif
+
 	/*
 	 * Setup response xdr_buf.
 	 * Initially it has just one page